### PR TITLE
[feature][segment] Support external meta V3 with common format

### PR DESCRIPTION
**Pull request by eldenmoon**
## Motivation

- Reduce `SegmentFooterPB.columns` bloat caused by Variant subcolumns and enable true on-demand meta loading.
- Introduce a general external column meta (CMO) layout that can be reused for non-Variant columns.
- Add `storage_format = V3` as “V2 + external column meta” to safely experiment in fuzzy tests.

## BE: External column meta (CMO) layout

Segment tail layout (not to scale):

```text
+------------------------------------------------------------------+
| ... data pages ...                                               |
|                                                                  |
| Column Meta Region (per top-level col, col_id ascending)        |
|   [Meta[col0]] [Meta[col1]] ... [Meta[colN-1]]                   |
|                                                                  |
| Footer (SegmentFooterPB)                                         |
|   col_meta_region_start      : uint64                            |
|   column_meta_entries        : repeated ColumnMetaEntryPB        |
|         - unique_id          : int32                             |
|         - length             : uint32                            |
|   num_columns                : uint32                            |
+------------------------------------------------------------------+
```

- Introduce `ExternalColMetaUtil` with:
  - `parse_external_meta_pointers()` to read `col_meta_region_start`, `column_meta_offsets_start`, `num_columns` from footer and validate them.
  - `parse_uid_to_colid_map()` to build a `uid -> col_id` map based on `top_level_column_uids[]`.
  - `is_valid_meta_slice()` to sanity-check `(pos, size)` against the external meta region with overflow guards.
  - `read_col_meta()` to read a single `ColumnMetaPB` by `col_id` from external meta region, using index cache lane.
  - `write_external_column_meta()` to:
    - externalize Variant subcolumns first (via `VariantExtMetaWriter`), then
    - write a contiguous Column Meta Region, then
    - write CMO entries, then
    - fill footer pointers and `top_level_column_uids`, finally clear `footer.columns()` to enable external-first loading.
- Wire this layout into `SegmentWriter` / `Segment` / `ColumnReader` so that:
  - top-level column metas can be loaded lazily via CMO,
  - old inline metas are no longer required once external meta is enabled.

## BE: Variant external meta (path-based subcolumn meta)

Variant subcolumn meta layout per root `uid`:

```text
variant_meta_keys.<uid>    : IndexedColumn<VARCHAR> (path)
  - value index: ON (fast seek_at_or_after by path)
  - ordinal index: OFF
  - encoding: PREFIX_ENCODING

footer.file_meta_datas:
  - key = "variant_meta_keys.<uid>"   , value = IndexedColumnMetaPB bytes
```

- Introduce `VariantExtMetaWriter` to preprocess `SegmentFooterPB.columns`:
  - For each column:
    - If no `column_path_info` → keep as top-level column.
    - If path relative to root is empty → treat as root Variant column and keep.
    - If relative path contains `__DORIS_VARIANT_SPARSE__` (including bucketized sparse cols like `.b{i}`) → collect as sparse subcolumns to be embedded into the root Variant’s `children_columns`.
    - Otherwise (non-sparse subcolumn) → serialize `ColumnMetaPB` and buffer as `(relative_path, meta_bytes)` under the root’s `uid`.
  - For each root `uid`, sort `(relative_path, meta_bytes)` by path and:
    - write keys (paths) into `variant_meta_keys.<uid>`,
    - write values (meta bytes) into `variant_meta_values.<uid>`,
    - persist the two `IndexedColumnMetaPB` as `file_meta_datas` entries.
  - Replace `footer.columns()` with:
    - only top-level columns (including Variant roots),
    - where each root has its sparse subcolumns merged into `children_columns`.
- Add `VariantExternalMetaReader` to read externalized Variant subcolumn metas:
  - `init_from_footer()`:
    - locate `variant_meta_keys.<uid>` / `variant_meta_values.<uid>` from `footer.file_meta_datas()`,
    - support legacy unsuffixed keys (`variant_meta_keys` / `variant_meta_values`) as a fallback,
    - open two `IndexedColumnReader`s and load their indexes.
  - `lookup_meta_by_path(rel_path)`:
    - seek on key column via value index,
    - use ordinal to read the matching value (meta bytes),
    - parse `ColumnMetaPB`.
  - `load_all()` / `load_all_once()`:
    - iterate the whole value column, parse each `ColumnMetaPB`,
    - reconstruct `SubcolumnColumnMetaInfo` tree by relative path,
    - accumulate `VariantStatistics` (e.g. `none_null_size`) when available.
  - `has_prefix(prefix)`:
    - perform `seek_at_or_after(prefix)` on keys and check `starts_with(prefix)` to answer “does any subcolumn with this path prefix exist?”.
- Update Variant `ColumnReader` / schema utilities to:
  - load root meta via CMO,
  - obtain non-sparse subcolumn metas via `VariantExternalMetaReader`,
  - obtain sparse/bucketized subcolumn metas from root’s `children_columns`,
  - keep compatibility with existing Variant formats while enabling V3 behavior when external meta is present.

## FE: V3 storage format and fuzz testing hooks

- In `PropertyAnalyzer`, extend `storage_format` parsing:
  - reject `"V1"` as deprecated and keep `"V2"` as default.
  - accept `"V3"` and map it to `TStorageFormat.V3` (semantically “V2 + external column meta support”).
- In `Config`:
  - add `random_use_v3_storage_format` (mutable, masterOnly) to allow FE to:
    - in fuzzy / regression tests, randomly choose some OLAP tables to use `storage_format = V3`,
    - increase coverage of V3 + ext-meta paths without impacting user-specified formats.
- In `CreateTableInfo`:
  - when engine is `OLAP` and `Config.random_use_v3_storage_format == true`,
    - if user does not explicitly set `storage_format`,
    - randomly set `storage_format = "V3"` for some tables and log this choice for observability.

## Compatibility & safety

- Existing segments without external meta remain readable:
  - CMO pointers validation will fail gracefully and readers fall back to inline footer metas.
  - Variant external meta reader stays in “unavailable” state when no `variant_meta_*` entries are present.
- All external meta slices are validated via `is_valid_meta_slice()` with overflow checks and region bounds.
- Footer write order ensures crash-consistency: no visible CMO pointers in footer are written before the underlying regions are fully flushed.

## Testing

- Add BE unit tests:
  - `external_col_meta_util_test` to exercise pointer parsing, CMO range validation and read-back correctness.
  - extend `variant_column_writer_reader_test` to cover V3 external meta for Variant, including sparse columns and path resolution.
  - extend `schema_util_rowset_test` to ensure schema evolution and rowset-level utilities understand external meta layout.
- Add regression tests under `variant_p0`:
  - `ext_meta/test_storage_format_v2_1` and multiple `test_variant_external_meta_*` suites (concurrent, edge cases, integration, with sparse),
  - `query_subcolumns` cases to verify subcolumn discovery, prefix checks and query correctness on V3 tables.
- Wire FE regression configs (`Config.groovy`, load scripts) to optionally enable `random_use_v3_storage_format` so fuzzy runs continuously hit V3 + ext-meta paths.

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Release note

None

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [ ] No.
    - [ ] Yes. 

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

**Comment by hello-stephen** [comment]
# FE UT Coverage Report
Increment line coverage `` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57702_8820c7c526f824f6b58d621b181e761357f37bf8/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57702_8820c7c526f824f6b58d621b181e761357f37bf8/fe_report/index.html)

**Comment by eldenmoon** [comment]
run p0 10

**Comment by eldenmoon** [comment]
run p0 10

**Comment by hello-stephen** [comment]
# BE Regression && UT Coverage Report
Increment line coverage `86.97% (681/783)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57702_8820c7c526f824f6b58d621b181e761357f37bf8_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57702_8820c7c526f824f6b58d621b181e761357f37bf8_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.68% (24398/34037) |
| Line Coverage     | 58.20% (254142/436644) |
| Region Coverage   | 53.74% (212670/395763) |
| Branch Coverage   | 54.95% (90653/164967) |

**Comment by eldenmoon** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/variant/variant_ext_meta_writer.cpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+Status VariantExtMetaWriter::externalize_from_footer(SegmentFooterPB* footer) {
+    // Variant meta pre-processing:
+    // - Collect non-sparse subcolumns per root and externalize them as KV (path→ColumnMetaPB bytes)
+    //   in sorted path order to keep stability;
+    // - Collect sparse subcolumns (including buckets) per root and embed them into the corresponding
+    //   root variant's ColumnMetaPB.children_columns;
+    // - Keep only top-level columns (including variant roots) in footer.columns().
+    std::vector<ColumnMetaPB> kept;
```
Comment (csun5285) [inline be/src/olap/rowset/segment_v2/variant/variant_ext_meta_writer.cpp]: Variant 的 root 列的ColumnMetaPB放在footer 里面，其他的子列的，稀疏列的ColumnMetaPB可以用一个单独的结构保存在内存里，这里就不用再动 footer 了。

File: be/src/olap/rowset/segment_v2/vertical_segment_writer.cpp [hunk 2]
```diff
@@ -1344,6 +1345,17 @@ Status VerticalSegmentWriter::_write_primary_key_index() {
 Status VerticalSegmentWriter::_write_footer() {
     _footer.set_num_rows(_row_count);
 
+    // Decide whether to externalize ColumnMetaPB by tablet default, and stamp footer version
+    DCHECK(_tablet != nullptr);
+    const bool enable_external_col_meta =
+            _tablet->tablet_meta()->is_external_segment_meta_used_default();
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/vertical_segment_writer.cpp]: 我们现在关于适用segmentv2还是v3 需要把整个tablet传递下来吗？

Comment (eldenmoon) [reply be/src/olap/rowset/segment_v2/vertical_segment_writer.cpp]: 或者放到tablet schema中

File: be/src/olap/tablet_meta.cpp [hunk 3]
```diff
@@ -149,10 +152,12 @@ TabletMeta::TabletMeta(int64_t table_id, int64_t partition_id, int64_t tablet_id
...
                        int64_t time_series_compaction_level_threshold,
                        TInvertedIndexFileStorageFormat::type inverted_index_file_storage_format,
-                       TEncryptionAlgorithm::type tde_algorithm)
+                       TEncryptionAlgorithm::type tde_algorithm,
+                       bool is_external_segment_meta_used_default)
         : _tablet_uid(0, 0),
           _schema(new TabletSchema),
-          _delete_bitmap(new DeleteBitmap(tablet_id)) {
+          _delete_bitmap(new DeleteBitmap(tablet_id)),
+          _is_external_segment_meta_used_default(is_external_segment_meta_used_default) {
```
Comment (yiguolei) [inline be/src/olap/tablet_meta.cpp]: _is_external_segment_meta_used_default 这个字段，跟我们的segmentv3 字段是什么关系？

Comment (eldenmoon) [reply be/src/olap/tablet_meta.cpp]: "storage_format" = "V3" 开启后 _is_external_segment_meta_used_default 在创建tablet的时候会设置成true

**Comment by hello-stephen** [comment]
# FE UT Coverage Report
Increment line coverage `  16.67% (2/12)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57702_22c56acb6fdbb080ac3ef44a829c31195df11696/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57702_22c56acb6fdbb080ac3ef44a829c31195df11696/fe_report/index.html)

**Comment by hello-stephen** [comment]
# FE UT Coverage Report
Increment line coverage `  16.67% (2/12)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57702_099689b8fa5843d33df45eb0d51b13041762f49f/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57702_099689b8fa5843d33df45eb0d51b13041762f49f/fe_report/index.html)

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 4]
```diff
@@ -248,12 +250,31 @@ Status ColumnReader::create(const ColumnReaderOptions& opts, const SegmentFooter
                             uint32_t column_id, uint64_t num_rows,
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/column_reader.cpp]: 我们之前这个接口有点失误，我们不应该传递footer，应该在外部把variant的subcolumns 这些解析出来，然后构建reader。
这里如果列真的很多，我们parse_external_meta_pointers 这个函数的计算复杂度还是很高的。
如果是在segment 里，我们就可以一把计算好，每个column metapb 在segment 文件里的offset和length，下次查询就只要在map 里找就可以了。

File: be/src/olap/rowset/segment_v2/segment.cpp [hunk 5]
```diff
@@ -681,15 +692,39 @@ Status Segment::_create_column_meta_once(OlapReaderStatistics* stats) {
...
+            }
+            _column_uid_to_footer_ordinal.try_emplace(column_meta.unique_id(), ordinal++);
         }
-        _column_uid_to_footer_ordinal.try_emplace(column_meta.unique_id(), ordinal++);
+        loaded = true;
+    }
+    // 3) If still not loaded, warn once to aid troubleshooting (damaged mapping and empty inline columns).
+    if (!loaded) {
+        LOG(WARNING) << "segment external meta mapping missing or corrupted and no inline columns; "
+                        "uid->col_id cannot be resolved. path="
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment.cpp]: 此时，为什么不直接报错？

File: be/src/olap/rowset/segment_v2/column_reader_cache.cpp [hunk 6]
```diff
@@ -155,31 +170,58 @@ Status ColumnReaderCache::get_path_column_reader(uint32_t col_uid,
...
-        // lazy create column reader from footer
-        DCHECK_GE(node->data.footer_ordinal, 0);
-        std::shared_ptr<SegmentFooterPB> footer_pb_shared;
-        {
-            std::lock_guard<std::mutex> lock(_cache_mutex);
-            // keep the lock until the footer is loaded, since _get_segment_footer is not thread safe
-            RETURN_IF_ERROR(_segment->_get_segment_footer(footer_pb_shared, stats));
-        }
-        const auto& col_footer_pb = footer_pb_shared->columns(node->data.footer_ordinal);
+
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/column_reader_cache.cpp:162]: 这里直接把get_path_column_reader 这个函数分成2个函数吧，一个 get_path_column_reader_v2 另外一个get_path_column_reader_v3

**Comment by hello-stephen** [comment]
# FE Regression Coverage Report
Increment line coverage `  28.57% (4/14)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57702_946eb74ee832f0fe4402a686dc53037cea686521_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57702_946eb74ee832f0fe4402a686dc53037cea686521_merge_fe/report/index.html)

File: be/src/olap/rowset/segment_v2/external_col_meta_util.cpp [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+namespace doris::segment_v2 {
+
+bool ExternalColMetaUtil::parse_external_meta_pointers(
+        const SegmentFooterPB& footer, ExternalColMetaUtil::ExternalMetaPointers* out) {
+    out->region_start = footer.col_meta_region_start();
+    out->num_columns = footer.num_columns();
+    // Validate pointers and basic consistency to ensure external meta is actually present.
+    if (out->region_start == 0) {
+        return false;
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/external_col_meta_util.cpp]: 不要返回true or false，返回error status，因为error status 里我们可以详细记录错误原因。 返回true or false，这部分信息就丢了，我们如果真的有bug，定位起来就很难了。

File: be/src/olap/rowset/segment_v2/column_meta_accessor.cpp [hunk 8]
```diff
@@ -0,0 +1,222 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/column_meta_accessor.cpp:2]: 我们需要为v2和v3 建立单测，否则后续的bugfix 和改进，都不会有单测去保证改动合理并且正确。

File: gensrc/proto/segment_v2.proto [hunk 9]
```diff
@@ -246,6 +256,17 @@ message SegmentFooterPB {
...
+    // Externalized ColumnMetaPB layout (optional; presence indicates v2 meta layout available)
+    // Start offset of the contiguous per-column ColumnMetaPB region (absolute file offset).
+    optional uint64 col_meta_region_start = 11;
+    // Per-column external meta entries, in col_id order. Each entry contains the top-level
+    // unique id and the serialized length (in bytes) of its ColumnMetaPB in the external
+    // Column Meta Region.
+    repeated ColumnMetaEntryPB column_meta_entries = 12;
+    // Number of top-level columns in this segment for external meta addressing (col_id in [0, N)).
+    // Must equal column_meta_entries.size().
+    optional uint32 num_columns = 13;
```
Comment (yiguolei) [inline gensrc/proto/segment_v2.proto]: 看注释num columns 一定跟column_meta_entries.SIZE 相等，我们保存他的意义是什么？

File: be/src/olap/rowset/segment_v2/external_col_meta_util.cpp [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        region_size += len;
+    }
+    out->region_end = out->region_start + region_size;
+    if (out->region_end <= out->region_start) {
+        return false;
+    }
+    return true;
+}
+
+bool ExternalColMetaUtil::parse_uid_to_colid_map(
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/external_col_meta_util.cpp]: 这里也不要返回true or false

File: be/src/olap/rowset/segment_v2/segment.cpp [hunk 11]
```diff
@@ -681,16 +635,12 @@ Status Segment::_create_column_meta_once(OlapReaderStatistics* stats) {
...
-            continue;
-        }
-        _column_uid_to_footer_ordinal.try_emplace(column_meta.unique_id(), ordinal++);
-    }
+    // Build uid -> col_id mapping used to locate per-column meta.
+    // The concrete layout (inline vs external CMO) is handled by ColumnMetaAccessor.
+    _column_uid_to_footer_ordinal.clear();
+    ColumnMetaAccessor accessor;
+    RETURN_IF_ERROR(accessor.init(footer, _file_reader));
+    RETURN_IF_ERROR(accessor.build_uid_to_colid_map(&_column_uid_to_footer_ordinal));
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment.cpp]: 我们到底是叫colid还是ordinal 得统一了_column_uid_to_footer_ordinal

File: gensrc/proto/segment_v2.proto (hunk 9 above)
Comment (eldenmoon) [reply gensrc/proto/segment_v2.proto]: 本来是用来校验的， 实际用不上，我删了

File: be/src/olap/rowset/segment_v2/segment.cpp (hunk 11 above)
Comment (eldenmoon) [reply be/src/olap/rowset/segment_v2/segment.cpp]: 统一成_column_uid_to_column_id

File: be/src/olap/rowset/segment_v2/column_meta_accessor.h [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    ~ColumnMetaAccessor();
+
+    // Initialize accessor and choose V2 / V3 implementation based on footer.
+    Status init(const SegmentFooterPB& footer, io::FileReaderSPtr file_reader);
+
+    // Build uid -> col_id mapping.
+    Status build_uid_to_colid_map(std::unordered_map<int32_t, size_t>* uid_to_colid) const;
+
+    // Get ColumnMetaPB for the given column_id.
+    Status get_column_meta(uint32_t column_id, ColumnMetaPB* out) const;
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/column_meta_accessor.h]: Status get_column_meta_by_uniqueid(uint32_t column_unique_id, ColumnMetaPB* out) const;
Status get_column_meta_by_id(uint32_t column_ordinal, ColumnMetaPB* out) const;

File: be/src/olap/rowset/segment_v2/column_meta_accessor.cpp [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+Status ColumnMetaAccessor::init(const SegmentFooterPB& footer, io::FileReaderSPtr file_reader) {
+    // First check footer version to see if external Column Meta Region might exist,
+    // then try to parse external layout; if parsing fails, fall back to V2.
+    ExternalColMetaUtil::ExternalMetaPointers ptrs;
+    if (footer.version() >= kSegmentFooterVersionV3_ExtColMeta &&
+        ExternalColMetaUtil::parse_external_meta_pointers(footer, &ptrs) && ptrs.num_columns > 0) {
+        _impl = new ColumnMetaAccessorV3(&footer, std::move(file_reader), ptrs);
+    } else {
+        _impl = new ColumnMetaAccessorV2(&footer, std::move(file_reader));
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/column_meta_accessor.cpp]: 这个直接用unique ptr吧

File: be/src/olap/rowset/segment_v2/segment.cpp (hunk 11 above)
Comment (eldenmoon) [reply be/src/olap/rowset/segment_v2/segment.cpp]: done

File: be/src/olap/rowset/segment_v2/external_col_meta_util.cpp (hunk 10 above)
Comment (eldenmoon) [reply be/src/olap/rowset/segment_v2/external_col_meta_util.cpp]: done

File: gensrc/proto/segment_v2.proto (hunk 9 above)
Comment (eldenmoon) [reply gensrc/proto/segment_v2.proto]: done

File: be/src/olap/rowset/segment_v2/column_meta_accessor.cpp (hunk 8 above)
Comment (eldenmoon) [reply be/src/olap/rowset/segment_v2/column_meta_accessor.cpp:2]: done

File: be/src/olap/rowset/segment_v2/column_meta_accessor.h (hunk 12 above)
Comment (eldenmoon) [reply be/src/olap/rowset/segment_v2/column_meta_accessor.h]: done

File: be/src/olap/rowset/segment_v2/column_meta_accessor.cpp (hunk 13 above)
Comment (eldenmoon) [reply be/src/olap/rowset/segment_v2/column_meta_accessor.cpp]: done

**Comment by hello-stephen** [comment]
# FE Regression Coverage Report
Increment line coverage `  28.57% (4/14)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57702_839366f500bab532082fcf3f97cf445bdb1e85b2_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57702_839366f500bab532082fcf3f97cf445bdb1e85b2_merge_fe/report/index.html)

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by hello-stephen** [comment]
# FE Regression Coverage Report
Increment line coverage `  44.44% (8/18)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57702_d3845d1c83423ecb203afaae91b10c0f950d78f8_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57702_d3845d1c83423ecb203afaae91b10c0f950d78f8_merge_fe/report/index.html)

**Comment by eldenmoon** [comment]
run p0 10

**Comment by hello-stephen** [comment]
# BE UT Coverage Report
Increment line coverage `84.45% (766/907)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/cc33be800f367dd823c6773f6272fec6bb00115d_cc33be800f367dd823c6773f6272fec6bb00115d/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/cc33be800f367dd823c6773f6272fec6bb00115d_cc33be800f367dd823c6773f6272fec6bb00115d/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 52.90% (18500/34969) |
| Line Coverage     | 38.46% (169504/440731) |
| Region Coverage   | 33.24% (131559/395776) |
| Branch Coverage   | 34.21% (56675/165649) |

**Comment by hello-stephen** [comment]
# FE Regression Coverage Report
Increment line coverage `  44.44% (8/18)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57702_cc33be800f367dd823c6773f6272fec6bb00115d_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57702_cc33be800f367dd823c6773f6272fec6bb00115d_merge_fe/report/index.html)

**Review by yiguolei — APPROVED** [review]

**Review by csun5285 — APPROVED** [review]
lgtm

File: be/src/olap/rowset/segment_v2/segment.h [hunk 14]
```diff
@@ -65,12 +65,19 @@ struct RowLocation;
 
 namespace segment_v2 {
 
+// SegmentFooterPB versioning:
+// 1 = V2 baseline
+// 2 = V3 (externalized ColumnMetaPB region + CMO present)
+static constexpr uint32_t kSegmentFooterVersionV2 = 1;
+static constexpr uint32_t kSegmentFooterVersionV3_ExtColMeta = 2;
```
Comment (gavinchou) [inline be/src/olap/rowset/segment_v2/segment.h:72]: make it pb

File: gensrc/proto/olap_file.proto [hunk 15]
```diff
@@ -484,6 +487,9 @@ message TabletSchemaCloudPB {
     optional int64 storage_page_size = 31 [default=65536];
     optional int64 storage_dict_page_size = 32 [default=262144];
 
+    // Whether new segments of this tablet use externalized ColumnMetaPB layout (CMO)
+    optional bool is_external_segment_meta_used_default = 33 [default = false];
```
Comment (gavinchou) [inline gensrc/proto/olap_file.proto:491]: naming

**Review by gavinchou — APPROVED** [review]
